File: class/class11.1/src/BookRecommendation.cpp
#include "BookRecommendation.hpp"

#include <chrono>
#include <algorithm>
#include <stdexcept>

BookRecommendation::BookRecommendation() {
    books = {
        "The Topographer’s Clown",
        "The Chamber of Beaver",
        "The Ironer of Kanban",
        "The Piglet of Tire",
        "The Border of the Unix",
        "The Half-Time Convince",
        "The Earthly Pillows",
        "The Censorship of the Ping",
        "The True Powers",
        "The Overturn of the Ling"
    };
    generator.seed(std::chrono::system_clock::now().time_since_epoch().count());
}

std::string BookRecommendation::getRecommendation() {
    if (books.empty()) {
        throw std::runtime_error("Adios amigo!");
    }
    std::uniform_int_distribution<int> distribution(0, books.size() - 1);
    int index = distribution(generator);
    return books[index];
}

void BookRecommendation::removeBook(const std::string& book) {
    books.erase(std::remove(books.begin(), books.end(), book), books.end());
}

File: home/home11.1/book_recommendation/src/book_recommendation.rs
use rand::seq::SliceRandom;
use rand::thread_rng;

pub struct BookRecommendation {
    books: Vec<&'static str>,
}

impl BookRecommendation {
    pub fn new() -> Self {
        Self {
            books: vec![
                "The Topographer’s Clown",
                "The Chamber of Beaver",
                "The Ironer of Kanban",
                "The Piglet of Tire",
                "The Border of the Unix",
                "The Half-Time Convince",
                "The Earthly Pillows",
                "The Censorship of the Ping",
                "The True Powers",
                "The Overturn of the Ling",
            ],
        }
    }

    pub fn get_recommendation(&mut self) -> Option<&'static str> {
        if self.books.is_empty() {
            None
        } else {
            let mut rng = thread_rng();
            Some(*self.books.choose(&mut rng).unwrap())
        }
    }

    pub fn remove_book(&mut self, book: &str) {
        self.books.retain(|&b| b != book);
    }
}
